### PR TITLE
YALB-937: Views: Filter by tag

### PR DESCRIPTION
**Pull request by marcb4k**
### Description of work
- Adds the ability to filter views basic lists by a taxonomy term in the tags vocabulary
- At the moment only one term is supported and help text is added to inform the content editor
- There is some work in Drupal core over single value autocomplete widgets not warning about cardinality: https://www.drupal.org/project/drupal/issues/2951134 so for the moment, if there are multiple values, the last value is used.

### Functional testing steps:
- [ ] Create a few News nodes and tag them with some tags. Make sure to reuse a tag or two for testing the view.
- [ ] Create a page and add a Views Basic paragraph to it
- [ ] Select "Display News articles as News cards"
- [ ] Start typing the name of one of the tags used above for autocomplete and select that tag
- [ ] Save the paragraph and verify that the preview adds the "Filtered by tag "{tagname}"
- [ ] Save the page
- [ ] Verify that the view displays only the news nodes tagged with the tag selected.

### CODE REVIEW

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/Plugin/Field/FieldWidget/ViewsBasicDefaultWidget.php [hunk 1]
```diff
@@ -156,6 +161,24 @@ public function formElement(
       '#suffix' => '</div>',
     ];
 
+    $form['group_user_selection']['markup_post_entity_types'] = [
+      '#type' => 'markup',
+      '#markup' => '</div>',
+    ];
+
+    // @todo add validation for only one term.
```
Comment (nJim) [inline web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/Plugin/Field/FieldWidget/ViewsBasicDefaultWidget.php:169]: Let's leave this for now as I want to keep charing forward. But if this becomes and issue we can change this to a text input and create our own autocomplete callback. The entity_autocomplete type is doing all of the work for us, but it sucks that it does not have enough options.

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/Plugin/views/filter/ViewsBasicFilter.php [hunk 2]
```diff
@@ -24,10 +24,15 @@ public function query() {
     $query = $this->query;
 
     // Parse content type filters.
-    // @todo Add taxonomy term as a filter. Code here: https://gist.github.com/marcb4k/fd4716eb133aa846f90a9e42c80def8c.
     foreach ($this->value['filters']['types'] as $content_type) {
       $query->addWhere($this->options['group'], 'type', $content_type);
     }
+
+    // Parse tag filters.
+    $tag_table = $query->addTable('node__field_tags');
```
Comment (nJim) [inline web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/Plugin/views/filter/ViewsBasicFilter.php:32]: WHAT? How did I not know `addTable` exists? I've been writing joins for years like a chump.

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php [hunk 3]
```diff
@@ -108,6 +108,26 @@ public function viewModeList($content_type) {
     return $viewModes;
   }
 
+  /**
+   * Returns an array of taxonomy tags.
+   */
+  public function tagsList() {
```
Comment (nJim) [inline web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php:114]: I don't think this method is in use. Should we delete it? or do you have something planned for it?

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php [hunk 4]
```diff
@@ -122,11 +142,20 @@ public function getViewModeLabel($type, $view_mode) {
     return self::ALLOWED_ENTITIES[$type]['view_modes'][$view_mode];
   }
 
+  /**
+   * Returns a tag label given a term ID.
+   */
+  public function getTagLabel($tag) {
+    $term = $this->entityTypeManager()->getStorage('taxonomy_term')->load($tag);
+    return $term->name->value;
```
Comment (nJim) [inline web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php:150]: $term may be null if the term does not exist. For example, if someone deletes the taxonomy term. Then I think these node forms will white-screen when trying to generate previews.

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php [hunk 5]
```diff
@@ -136,6 +165,11 @@ public function getDefaultParamValue($type, $params) {
         $defaultParam = $paramsDecoded['filters']['types'][0];
         break;
 
+      case 'tags':
+        $tid = (int) $paramsDecoded['filters']['tags'][0];
```
Comment (nJim) [inline web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php:169]: I have a feeling that this method will change a lot in the near future as we make the data more complex. So I don't want to overthinking it. 

The only think we may want to change right now is avoiding warnings around `$paramsDecoded['filters']['types'][0]`. It is possible that these keys will not exist. (Okay it is not really possible since this tool is not live, but I'm approaching this from a future-proof point of view.) So we may want to check if !empty($paramsDecoded['filters']['types'][0]) to avoid "illegal offset" and "Undefined array key" warnings.

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php [hunk 6]
```diff
@@ -122,11 +142,20 @@ public function getViewModeLabel($type, $view_mode) {
     return self::ALLOWED_ENTITIES[$type]['view_modes'][$view_mode];
   }
 
+  /**
+   * Returns a tag label given a term ID.
+   */
+  public function getTagLabel($tag) {
```
Comment (nJim) [inline web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php:148]: I didn't think of this before, but would you mind typehinting functions? For example:
```
/**
  * Returns a tag label given a term ID.
  * 
  * @param int $tag
  *   The ID of a taxonomy term.
  * 
  * @return string
  *   The label of the given taxonomy term or an empty string.
  */
  public function getTagLabel(int $tag): string {
```

File: web/profiles/custom/yalesites_profile/composer.json [hunk 7]
```diff
@@ -89,6 +89,9 @@
       },
       "drupal/core": {
         "plural results summary https://www.drupal.org/project/drupal/issues/2888320": "https://www.drupal.org/files/issues/2021-12-15/2888320-78.patch"
+      },
+      "drupal/hide_revision_field": {
+        "fix deprecated function unserialize https://www.drupal.org/project/hide_revision_field/issues/3260101": "https://www.drupal.org/files/issues/2022-01-24/3260101-2.patch"
```
Comment (nJim) [inline web/profiles/custom/yalesites_profile/composer.json:94]: Is this related?

**Review by nJim — CHANGES_REQUESTED** [review]
Everything works great. I added some commentary to parts of the code. A few changes requested that are mostly from an opinionated place.

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/Plugin/Field/FieldWidget/ViewsBasicDefaultWidget.php (hunk 1 above)
Comment (marcb4k) [reply web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/Plugin/Field/FieldWidget/ViewsBasicDefaultWidget.php:169]: Yah, from that discussion on d.o, it seems that work is progressing. We could try applying that patch if you'd like?

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/Plugin/views/filter/ViewsBasicFilter.php (hunk 2 above)
Comment (marcb4k) [reply web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/Plugin/views/filter/ViewsBasicFilter.php:32]: It had been a while since I've done queries like that, so I looked it up and saw that - pretty neat! I recall doing joins before as well.

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php (hunk 3 above)
Comment (marcb4k) [reply web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php:114]: Good catch - before I was using the `entity_autocomplete`, I was going to write my own. Removed.

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php (hunk 5 above)
Comment (marcb4k) [reply web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php:169]: Good call. Fixed.

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php (hunk 6 above)
Comment (marcb4k) [reply web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/ViewsBasicManager.php:148]: Yup, thanks for the reminder. Added for all functions.

File: web/profiles/custom/yalesites_profile/composer.json (hunk 7 above)
Comment (marcb4k) [reply web/profiles/custom/yalesites_profile/composer.json:94]: Not to this work, but I received an error about this because of our upgraded PHP version. I can remove this and add to a different PR if you'd like?

Comment (nJim) [reply web/profiles/custom/yalesites_profile/composer.json:94]: Nope it's good. Just wan't sure.

File: web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/Plugin/Field/FieldWidget/ViewsBasicDefaultWidget.php (hunk 1 above)
Comment (nJim) [reply web/profiles/custom/yalesites_profile/modules/custom/ys_views_basic/src/Plugin/Field/FieldWidget/ViewsBasicDefaultWidget.php:169]: It's good for now. We just want something stable enough to get client feedback. We'll just keep things moving for now.

**Review by nJim — APPROVED** [review]
Approving change. Thanks for all the little and big things in this one.
